// @chenglou/pretext/dist/layout.js
const measureCanvas = document.createElement('canvas');
const measureContext = measureCanvas.getContext('2d');

export function prepareWithSegments(value, font) {
  measureContext.font = font;
  const chars = Array.from(String(value));
  return {
    segments: chars,
    widths: chars.map((char) => measureContext.measureText(char).width)
  };
}

(function installHorseAsciiRepair() {
  const COLS = 60;
  const ROWS = 37;
  const CW = 7;
  const CH = 10;
  const FILL_CHAR = '.';

  function ready(callback) {
    if (document.readyState === 'loading') {
      document.addEventListener('DOMContentLoaded', callback, { once: true });
    } else {
      callback();
    }
  }

  function buildGrid(img) {
    const sw = COLS * 2;
    const sh = ROWS * 2;
    const tmp = document.createElement('canvas');
    tmp.width = sw;
    tmp.height = sh;
    const ctx = tmp.getContext('2d', { willReadFrequently: true });
    ctx.fillStyle = '#fff';
    ctx.fillRect(0, 0, sw, sh);
    ctx.drawImage(img, 0, 0, sw, sh);
    const px = ctx.getImageData(0, 0, sw, sh).data;
    const lum2 = new Float32Array(sw * sh);

    for (let i = 0; i < sw * sh; i++) {
      lum2[i] = (px[i * 4] * 0.299 + px[i * 4 + 1] * 0.587 + px[i * 4 + 2] * 0.114) / 255;
    }

    const grid = new Float32Array(COLS * ROWS);
    for (let r = 0; r < ROWS; r++) {
      for (let c = 0; c < COLS; c++) {
        const r2 = r * 2;
        const c2 = c * 2;
        grid[r * COLS + c] = Math.min(
          lum2[r2 * sw + c2],
          lum2[r2 * sw + c2 + 1],
          lum2[(r2 + 1) * sw + c2],
          lum2[(r2 + 1) * sw + c2 + 1]
        );
      }
    }
    return grid;
  }

  function buildSoftInterior(grid) {
    const mask = new Uint8Array(COLS * ROWS);
    for (let r = 0; r < ROWS; r++) {
      let first = -1;
      let last = -1;
      let darkCount = 0;
      for (let c = 0; c < COLS; c++) {
        if (grid[r * COLS + c] < 0.72) {
          if (first < 0) first = c;
          last = c;
          darkCount++;
        }
      }
      if (first < 0 || last - first < 4 || darkCount < 3) continue;
      const pad = r > 7 && r < 29 ? 1 : 0;
      for (let c = Math.max(0, first - pad); c <= Math.min(COLS - 1, last + pad); c++) {
        if (grid[r * COLS + c] > 0.74) mask[r * COLS + c] = 1;
      }
    }
    return mask;
  }

  ready(() => {
    const canvas = document.getElementById('horse-canvas');
    const img = document.getElementById('img-horse');
    if (!canvas || !img) return;

    const ctx = canvas.getContext('2d');
    let mask = null;

    function prepare() {
      if (!img.naturalWidth) return false;
      mask = buildSoftInterior(buildGrid(img));
      return true;
    }

    if (!prepare()) {
      img.addEventListener('load', prepare, { once: true });
    }

    function drawRepair() {
      if (mask) {
        const isLight = document.documentElement.dataset.theme === 'light';
        ctx.save();
        ctx.textBaseline = 'alphabetic';
        ctx.font = `bold ${CH}px 'Courier New', monospace`;
        ctx.fillStyle = isLight ? 'rgba(0,0,0,0.12)' : 'rgba(255,255,255,0.13)';
        for (let r = 0; r < ROWS; r++) {
          for (let c = 0; c < COLS; c++) {
            if (mask[r * COLS + c]) ctx.fillText(FILL_CHAR, c * CW, (r + 1) * CH);
          }
        }
        ctx.restore();
      }
      requestAnimationFrame(drawRepair);
    }

    requestAnimationFrame(drawRepair);
  });
})();
